mm2src/common/custom_futures/repeatable/mod.rs

```diff
@@ -0,0 +1,313 @@
+/// * `Ok(T)` => `Action::Ready(T)`
+/// * `Err(E)` => `Action::Retry(E)`
+#[macro_export]
+macro_rules! retry_on_err {
+    (async { $($t:tt)* }) => {
+        $crate::custom_futures::repeatable::Repeatable::new(|| {
+            use $crate::custom_futures::repeatable::RetryOnError;
+            use futures::FutureExt;
+
+            let fut = async { $($t)* };
+            Box::pin(fut.map(Result::retry_on_err))
+        })
+    };
+    ($fut:expr) => {
+        $crate::custom_futures::repeatable::Repeatable::new(|| {
+            use $crate::custom_futures::repeatable::RetryOnError;
+            use futures::FutureExt;
+
+            $fut.map(Result::retry_on_err)
+        })
+    };
+}
+
+/// The macro expands as `return Action::Ready(T)`.
+#[macro_export]
+macro_rules! ready {
+    () => {{
+        return $crate::custom_futures::repeatable::Action::Ready(());
+    }};
+    ($res:expr) => {{
+        return $crate::custom_futures::repeatable::Action::Ready($res);
+    }};
+}
+
+/// The macro expands as `return Action::Retry(E)`.
+#[macro_export]
+macro_rules! retry {
+    () => {{
+        return $crate::custom_futures::repeatable::Action::Retry(());
+    }};
+    ($err:expr) => {{
+        return $crate::custom_futures::repeatable::Action::Retry($err);
+    }};
+}
+
+/// Unwraps a result or returns `Action::Retry(E)`.
+#[macro_export]
+macro_rules! try_or_retry {
+    ($exp:expr) => {{
+        match $exp {
+            Ok(t) => t,
+            Err(e) => $crate::retry!(e),
+        }
+    }};
+}
+
+/// Unwraps a result or returns `Action::Ready(E)`.
+#[macro_export]
+macro_rules! try_or_ready_err {
+    ($exp:expr) => {{
+        match $exp {
+            Ok(t) => t,
+            Err(e) => $crate::ready!(Err(e)),
+        }
+    }};
+}
+
+pub trait FactoryTrait<F>: Fn() -> F {}
+
+impl<Factory, F> FactoryTrait<F> for Factory where Factory: Fn() -> F {}
+
+pub trait RepeatableTrait<T, E>: Future<Output = Action<T, E>> + Unpin {}
+
+impl<F, T, E> RepeatableTrait<T, E> for F where F: Future<Output = Action<T, E>> + Unpin {}
+
+pub(crate) trait InspectErrorTrait<E>: 'static + Fn(&E) + Send {}
+
+impl<F: 'static + Fn(&E) + Send, E> InspectErrorTrait<E> for F {}
+
+/// The future is ether ready (with a `T` result), or not ready (failed with an intermediate `E` error).
+#[derive(Debug)]
+pub enum Action<T, E> {
+    Ready(T),
+    Retry(E),
+}
+
+pub trait RetryOnError<T, E> {
+    fn retry_on_err(self) -> Action<T, E>;
+}
+
+impl<T, E> RetryOnError<T, E> for Result<T, E> {
+    /// Converts `Result<T, E>` into `Action<T, E>`:
+    /// * `Ok(T)` => `Action::Ready(T)`.
+    /// * `Err(E)` => `Action::Retry(E)`.
+    #[inline]
+    fn retry_on_err(self) -> Action<T, E> {
+        match self {
+            Ok(ready) => Action::Ready(ready),
+            Err(e) => Action::Retry(e),
+        }
+    }
+}
+
+/// The result of `repeatable` or `retry_on_err` macros - the first step at the future configuration.
+pub struct Repeatable<Factory, F, T, E> {
```

Comment:
I think that `RepeatEvery`, `RepeatUntil` and `RepeatAttempts` functionalities can all be implemented using this `Repeatable` struct only, I guess this will require some optional parameters and an enum for either `RepeatUntil` or `RepeatAttempts`. There are also some duplication in the implementation of these structs that can be removed if this is done.

I guess you wanted to constrain the user to use `RepeatEvery` then either `RepeatUntil` or `RepeatAttempts`. Do you think there can be any cases were we would need to wait for a timeout/attempts in a loop without needing to sleep?

Reply:
Yes, I tried to limit the user to only two possible options: either Until or Attempts.
But I'd love to use `Repeatable` structure only. But we should decide how to deal with the following cases:

```rust
// 1 case:  Do we need to execute the following futures only once?
let res1 = repeatable!(async { retry!() }).await;
let res2 = repeatable!(async { retry!() }).repeat_every_secs(1.).await;

// 2 case: Do we need to retry the future for 10 seconds AND up to 10 times?
let res = repeatable!(async { retry!() }).repeat_every_secs(1.).with_timeout_secs(10.).attempts(10).await;

// 3 case: Do we need to retry the following future without any sleep?
let res = repeatable!(async { retry!() }).attempts(10).await;
```

Reply:
```rust
// 1 case:  Do we need to execute the following futures only once?
let res1 = repeatable!(async { retry!() }).await;
let res2 = repeatable!(async { retry!() }).repeat_every_secs(1.).await;
```
Since there is no timeout or attempts, I believe the above case should be repeated endlessly.

```rust
// 2 case: Do we need to retry the future for 10 seconds AND up to 10 times?
let res = repeatable!(async { retry!() }).repeat_every_secs(1.).with_timeout_secs(10.).attempts(10).await;
```
For this case, I believe it's intuitive to make the future ready when either 10 seconds pass or 10 attempts were tried.

```rust
// 3 case: Do we need to retry the following future without any sleep?
let res = repeatable!(async { retry!() }).attempts(10).await;
```
Yes, this should be retried without sleep.

I understand it's not ideal to allow the user to make mistakes like these (not allowing a future to sleep or endless retries), restricting the user might be better actually after thinking about this. For endless loops and no sleep the user can do those explicitly without using `repeatable!` macro.

Reply:
Someone can forget to specify `with_timeout` and `attempts`, and it will lead to an endless loop that can be recognised by clippy or the compiler.
> Since there is no timeout or attempts, I believe the above case should be repeated endlessly.

Reply:
Like I said, restricting the user might be better after thinking more about this. All the other cases are rarely needed, they can be done regularly using loops and not through this macro. I will resolve this comment if you agree :)

Reply:
I still think that it could be better to have only `Repeatable` future. @caglaryucekaya @artemii235 what do you think about Omar's suggestion and the cases that I explained above?

Reply:
> Someone can forget to specify `with_timeout` and `attempts`, and it will lead to an endless loop that can't be recognised by clippy or the compiler.
> 
> > Since there is no timeout or attempts, I believe the above case should be repeated endlessly.
> 
> What if we run the future only once in this case
> 
> ```rust
> let res1 = repeatable!(async { retry!() }).await;
> ```
> 
> But we can also add `until_ready` so the future will be repeated until it's ready:
> 
> ```rust
> let res1 = repeatable!(async { retry!() }).until_ready().await;
> ```
> 
> What do you think?

"_Someone can forget to specify with_timeout and attempts, and it will lead to an endless loop that can't be recognised by clippy or the compiler._"

I think I agree with @sergeyboyko0791  in this case and a doc! comment can be very useful about this usage..

Reply:
Using only the `Repeatable` struct can lead to errors that no one intended like running the loop only once or endlessly, but I think that can be traded-off to keep the code simpler and to offer more flexible functionality. Warning the users about these cases with documentation should be enough I guess.

Reply:
Great, thanks everyone! I'll combine `RepeatableUntil` and `RepeatableAttempts` structures into one `Repeatable`. And I stand for running the future only once in the following case:
```rust
let res1 = repeatable!(async { retry!() }).await;
```

Reply:
> Great, thanks everyone! I'll combine RepeatableUntil and RepeatableAttempts structures into one Repeatable

I agree now since it's 3 against 1 😅

Reply:
I read your conversation. Just want to say, that I like the final idea to have only Repeatable and run future only once too. Other ideas are also not ideal, so I prefer to keep the code simple.

Reply:
Done